pkg/sos/serviceFuncs.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package sos
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/gorilla/mux"
+)
+
+func RegistersNeccesaryPatterns(router *mux.Router) {
+	router.HandleFunc("/ping", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("pong"))
+	}).Methods("GET")
+}
+
+func RegisterServiceWithSos(service string, port uint) error {
+	return registerServiceWithSos(service, port, "http://localhost:"+PortNum)
+}
+
+func registerServiceWithSos(service string, port uint, sosServerURL string) error {
+	m := RegisterReqJson{service, port}
+	b, err := json.Marshal(m)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("POST", sosServerURL+"/register", bytes.NewBuffer(b))
+	if err != nil {
+		return err
+	}
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	decoder := json.NewDecoder(res.Body)
+	defer res.Body.Close()
+	var retMsg struct{ Error string }
+	if err := decoder.Decode(&retMsg); err != nil {
+		return err
+	}
+
+	if retMsg != struct{ Error string }{} {
+		return fmt.Errorf(retMsg.Error)
+	}
+
+	return nil
+}
+
+func UnregisterServiceWithSos(service string) error {
+	return unregisterServiceWithSos(service, "http://localhost:"+PortNum)
+}
+
+func unregisterServiceWithSos(service string, sosServerURL string) error {
+	m := UnRegisterReqJson{service}
+	b, err := json.Marshal(m)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("POST", sosServerURL+"/unregister", bytes.NewBuffer(b))
+	if err != nil {
+		return err
+	}
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	decoder := json.NewDecoder(res.Body)
+	defer res.Body.Close()
+	var retMsg struct{ Error string }
+	if err := decoder.Decode(&retMsg); err != nil {
+		return err
+	}
+
+	if retMsg != struct{ Error string }{} {
+		return fmt.Errorf(retMsg.Error)
+	}
+
+	return nil
+}
+
+func StartServiceServer(router *mux.Router, serviceName string, portNumReq *uint) error {
+	listener, err := net.Listen("tcp", "localhost:0")
+	defer listener.Close()
+	if err != nil {
+		return err
+	}
+	port, err := strconv.ParseUint(strings.Split(listener.Addr().String(), ":")[1], 10, 32)
```

Comment:
This is fragile if for any reason the listener.Addr().String() ever has less than 2 elements. Could that happen? You may want to do the split and check for, since if that can happen, we'll get mysterious panics.

Reply:
Fixed.

Technically, if we are using tcp/ip, the address should always be ip:port. But I separate them just in case we change the implementation in the future or if something changes with the API.